spec/connection.md

```diff
@@ -0,0 +1,229 @@
+3. If Server [INFO][INFO] contains `tls_required` set to `true`, or the client has a tls requirement set to `true`, the client performs a TLS upgrade.
+4. Client sends [CONNECT][CONNECT] json.
+5. Clients and Server start to exchange PING/PONG messages to detect if the connection is alive.
+
+##### TLS First
+
+This method has been available since NATS Server 2.11.
+
+There are two prerequisites to use this method:
+1. Server config has enabled `handshake_first` field in the `tls` block.
+2. The client has set the `tls_first` option set to true.
+
+**handshake_first**
+has those possible values:
+- **`false`**: handshake first is disabled. Default value
+- `true`: handshake first is enabled and enforced. Clients that do not use this flow will fail to connect.
+- `duration` (i.e. 2s): a hybrid mode that will wait a given time, allowing the client to follow the `tls_first` flow. After the duration has expired, `INFO` is sent, enabling standard client TLS flow.
+- `auto`: same as above, with some default value. By default it waits 50ms for TLS upgrade before sending the [INFO][INFO].
+
+The flow itself is flipped. TLS is established before the Server sends INFO:
+
+1. Clients initiate a TCP connection to the Server.
+2. Client upgrades the connection to TLS.
+2. Server [INFO][INFO] json.
+4. Client sends [CONNECT][CONNECT] json.
+5. Clients and Server start to exchange PING/PONG messages to detect if the connection is alive.
+
+
+### Servers discovery
+
+When Server sends back [INFO][INFO]. It may contain additional URLs to which the client can make connection attempts.
+The client should store those URLs and use them in the Reconnection Strategy.
+
+A client should have an option to turn off using advertised URLs.
+By default, those URLs are used.
+
+### Reconnection Strategies (In progress)
+
+#### Detecting disconnection
+
+There are two methods that clients should use to detect disconnections:
+1. Missing two consecutive PONGs from the Server.
+2. Handling errors from TCP connection.
+
+#### Reconnect process
+
+When the client detects disconnection, it starts to reconnect attempts with the following rules:
+1. Immediate reconnect attempt
+    The client attempts to reconnect immediately after finding out it has been disconnected.
+2. Exponential BackOff with Jitter
+   - When the first reconnect fails, the backoff process should kick in. Default Jitter should also be included to avoid thundering herd problems.
+3. If the Server returned additional URLs, the client should try reconnecting in random order to each Server on the list.
+4. Successful reconnect resets the timers
+5. Upon reconnection, clients should resubscribe to all created subscriptions.
+
+If there is any change in the connection state - connected/disconnected, the client should have some way of notifying the user about it.
+This can be a callback function or any other idiomatic mechanism in a given language for reporting asynchronous events.
+
+## Reference-level Explanation
```

Comment:
maybe worth covering settings that influence the ping/pong and buffer sizes as those are generally critical to expose